### PR TITLE
[virtual_nfts][5] Replacing usage of MerkleUniqueToken with an intermediate adapter to smooth code transition for virtualization nfts

### PR DESCRIPTION
**Pull request by rjuang**
This PR creates a UniqueTokenAdapter which then serves as a replacement for MerkleUniqueToken. The adapter class will be used as a drop-in to quickly switch between accessing the virtual vs non-virtualized NFT instance. This will enable guarding virtualizing NFTs behind a feature flag.

Related issue: #3089 

### Note for reviewers:

- Review the file: `UniqueTokenAdapter.java` first.
- All changes after that is to switch `MerkleUniqueToken` to `UniqueTokenAdapter`.

### CODE REVIEW

**Comment by rjuang** [comment]
looks like there were some merge issues.. will move this back to draft until resolved.

**Comment by rjuang** [comment]
Fixed merging issues.

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 1]
```diff
@@ -230,12 +232,25 @@ private void mapModelChanges(Token token, MerkleToken mutableToken) {
...
 
-    private void mapModelChanges(UniqueToken nft, MerkleUniqueToken mutableNft) {
-        mutableNft.setOwner(nft.getOwner().asEntityId());
-        mutableNft.setSpender(nft.getSpender().asEntityId());
-        mutableNft.setMetadata(nft.getMetadata());
-        final var creationTime = nft.getCreationTime();
-        mutableNft.setPackedCreationTime(
-                packedTime(creationTime.getSeconds(), creationTime.getNanos()));
+    private void mapModelChanges(UniqueToken nft, UniqueTokenAdapter mutableNft) {
+        if (mutableNft.isVirtual()) {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:236]: Is there a reason to not just use the `UniqueTokenAdapter` top-level methods here?

**Review by tinker-michaelj — APPROVED** [review]
LGTM! To switch to using virtual NFTs, we just need to:

1. Use the virtual entity factory in `StoresModule.provideNftsLedger()`; and,
2. Change the wrapped entity constructed in `TypedTokenStore.persistMinted()`
?

**Comment by rjuang** [comment]
> LGTM! To switch to using virtual NFTs, we just need to:
> 
> 1. Use the virtual entity factory in `StoresModule.provideNftsLedger()`; and,
> 2. Change the wrapped entity constructed in `TypedTokenStore.persistMinted()`
> 
> ?

Almost.  
1. We need to also swap out MerkleMap<EntityNumPair, MerkleUniqueToken> with a UniqueTokenMapAdapter (this is the next PR coming up).
2. Then create a provider for UniqueTokenMapAdapter that will either use a MerkleMap or VirtualMap backing.
3. And finally, integrate migration code so that MerkleMap can be converted to VirtualMap

**Comment by tinker-michaelj** [comment]
> > LGTM! To switch to using virtual NFTs, we just need to:
> > 
> > 1. Use the virtual entity factory in `StoresModule.provideNftsLedger()`; and,
> > 2. Change the wrapped entity constructed in `TypedTokenStore.persistMinted()`
> > 
> > ?
> 
> Almost.
> 
> 1. We need to also swap out MerkleMap<EntityNumPair, MerkleUniqueToken> with a UniqueTokenMapAdapter (this is the next PR coming up).
> 2. Then create a provider for UniqueTokenMapAdapter that will either use a MerkleMap or VirtualMap backing.
> 3. And finally, integrate migration code so that MerkleMap can be converted to VirtualMap

Nice. 👍

**Pull request merged**